### ISSUE TITLE
Begin(itemcollection) validation

### ISSUE THREAD

**Issue by RenkoQC** [JeremySkinner/FluentValidation#599]
Hi,

First of all, english is not my main language, hope my text won't be too hard to read.

I got PersonViewModel

```
public PersonViewModel(){
Adresse = new List<AdresseViewModel>();
}

public  int ID{get; set;}
public string Name{get;set;}
public virtual List<AdresseViewModel> Adresse {get; set;}
```

And a AdresseViewModel

```
public  int ID{get; set;}
public String theAdr{get;set;}
```

I create a validator for both of them

PersonViewModelValidator

        public IdentificationEntrepriseViewModelValidator()
        {
          RuleFor(x=>x.Name).NotEmpty().WithMessage(x => "* {PropertyName}  is required");
          RuleFor(x=>x.Adresse).SetCollectionValidator(new AdresseViewModelValidator);
        }

AdresseViewModelValidator
 ```
       public AdresseViewModelValidator()
        {
            RuleFor(x => x.theAdr).NotEmpty().WithMessage(x => "* {PropertyName}  is required");
       } 
```

The validation for the Person name work as plan but for the adresse, I don't see any message in the    @Html.ValidationMessageFor(x => x.theAdr)

Can it be because I use the BeginCollectionItem ?

My Adresse fields look like this

```
<span class="field-validation-valid" data-valmsg-replace="true" data-valmsg-for="Adresse[09eb68dc-7710-424e-aa99-832d503c9db5].theAdr"></span>
<label for="Adresse_09eb68dc-7710-424e-aa99-832d503c9db5__Adr">Adresse</label>
<input name="Adresse[09eb68dc-7710-424e-aa99-832d503c9db5].theAdr" class="form-control" id="Adresse_09eb68dc-7710-424e-aa99-832d503c9db5__theAdr" type="text" value="">
```

I don't get any message.

But if I don't put something in the theAdr field, the **TryValidateModel** for my IdentificationEntrepriseViewModel return false.

Also, I try this

```
            var validator = new IdentificationEntrepriseViewModelValidator();
            var results = validator.Validate(ievm);

            foreach (var result in results.Errors)
            {
                Console.WriteLine("Property name: " + result.PropertyName);
                Console.WriteLine("Error: " + result.ErrorMessage);
                Console.WriteLine("");
            }

```
And it's actually show every field who don't respect the validation of my AdresseViewModelValidation

Do you have any idea why the validation don't appear on my form after server-side validation ?

Thanks !

**Comment by JeremySkinner** [comment]
Hi,

FluentValidation does not work with BeginItemCollection I’m afraid. Th GUIDs generated by BeginItemCollection mess up the validation process because they don’t match the indices in the Adresse property (it’s a List, which uses numeric indices...these can’t be matched up with the GUID suffixes) 

For more info see this post : https://github.com/JeremySkinner/FluentValidation/issues/535

**Comment by RenkoQC** [comment]
And there is nothing I can do about it ?

It's the BeginItemCollection OR the fluentvalidation ? :(

**Comment by JeremySkinner** [comment]
There is no workaround I'm afraid. 

If BeginItemCollection used numeric indicies, it would work correctly as the index type for `List<T>` is always an integer. But by using a GUID, it's impossible to match the entries in ModelState with the inputs being generated as FluentValidation always associates the error message against the *actual* index in the collection. 

FluentValidation essentially generates ModelState entries like this:

```
Property: Adresse[0].theAdr Message: "* theAdr is required"
Property: Adresse[1].theAdr Message: "* theAdr is required"
Property: Adresse[3].theAdr Message: "* theAdr is required"
```
...etc

...but when you make a call to `Html.ValidationMessageFor(x => x.theAdr)` inside BeginCollectionItem, it'll will look in ModelState for a property called "Adresse[09eb68dc-7710-424e-aa99-832d503c9db5]" instead, which doesn't exist (because it's actually "Adresse[0].theAdr").  

Ideally, BeginCollectionItem should be clever enough to generate the index based on the correct type of the collection...but it doesn't, it always uses a guid.

**Comment by JeremySkinner** [comment]
One thing you could try is instead of using List<T>, would be to use a custom collection type that uses a Guid-based indexer. I have no idea if this would work as I've not tested it, but might be worth investigating.

**Comment by RenkoQC** [comment]
Will take a look ! Thanks a lot !
 
Took me a lots of time to make BIC to work on my project, I don't want to start everything again but I also love how FluentValidation work. It's simple and clean..... Damn :(

**Comment by RenkoQC** [comment]
I might have found a solution. It's not a cute one, but from the test I made ... it seem to work.

I add  a 

> String Index{get;set;}

 to my AdresseViewModel

This index his a hidden value which content the GUID of the the current model.
   
In my exemple, my GUID is made from the ViewData.TemplateInfo.HtmlFieldPrefix instead of a String.

> 
> @using (Html.BeginCollectionItem(ViewData.TemplateInfo.HtmlFieldPrefix)){
> 
>     @Html.HiddenFor(x=>x.index, new { Value = @ViewData.TemplateInfo.HtmlFieldPrefix })
> 
> }

On controller validation

 ```
           if (TryValidateModel(ievm))
            {
              // BE HAPPY
            }
            else{
                var validator = new IdentificationEntrepriseViewModelValidator();
                var results = validator.Validate(ievm);

                foreach (var result in results.Errors)
                {
                    String index = result.PropertyName.Split('[', ']')[1];
                    int indexInt = 0;

                    if(!Int32.TryParse(index, out indexInt)){
                        // TO STUFF
                    }
                    String property = result.PropertyName.Split('.')[1];

                    ModelState.AddModelError(ievm.Adresse[indexInt].index.ToString() + "."+propriete, result.ErrorMessage);
                }
            }
```

It's not a clean code ( and I'm pretty new to C# ) , and probably need more test/changes but it do what I want it to to do.

Basically, the property name got the index of the invalid element and with the index I got the correct GUID from the hidden field . With that I can manually set the ModelError.

Maybe it may help someone else with a similar situation.

**Comment by JeremySkinner** [comment]
Thanks for sharing your workaround, manipulating the modelstate isn’t pretty but seems like a reasonable solution in this case.

**Comment by RenkoQC** [comment]
One last question !

Is it normal that even client side doesnt work ?

With my workaround, the server-side validation work but I was hoping that client-side would still work even if I use BeginCollectionItem.

3/4 of my project is using BeginCollectionItem .... do I have to do javascript client-side validation by myself ?

**Comment by JeremySkinner** [comment]
Yes I'm pretty sure that they're not compatible
